larpandoracontent/LArControlFlow/BdtBeamParticleIdTool.h

```diff
@@ -0,0 +1,505 @@
+         *
+         *  @param  rhs the SliceFeatures to copy
+         */
+        SliceFeatures(const SliceFeatures &rhs);
+
+        /**
+         *  @brief  Assignment operator
+         *
+         *  @param  rhs the SliceFeatures to assign
+         */
+        SliceFeatures &operator=(const SliceFeatures &rhs);
+
+        /**
+         *  @brief  Destructor
+         */
+        ~SliceFeatures();
+
+        /**
+         *  @brief  Check if all features were calculable
+         *
+         *  @return true if the feature vector is available
+         */
+        bool IsFeatureVectorAvailable() const;
+
+        /**
+         *  @brief  Get the feature vector for the SVM
+         *
+         *  @param  featuresVector empty feature vector to populate
+         */
+        void GetFeatureVector(LArMvaHelper::MvaFeatureVector &featureVector) const;
+        
+        /**
+         *  @brief  Get the probability that this slice contains a beam particle 
+         *
+         *  @param  adaBoostDecisionTree the adaptive boost decision tree used to calculate the probability
+         *
+         *  @return the probability that the slice contains a beam particle
+         */
+        float GetAdaBoostDecisionTreeScore(const AdaBoostDecisionTree &adaBoostDecisionTree) const;
+
+    private:
+        /**
+         *  @brief  Select a given fraction of a slice's calo hits that are closest to the beam spot
+         *
+         *  @param  inputCaloHitList all calo hits in slice
+         *  @param  outputCaloHitList to receive the list of selected calo hits
+         *  @param  closestHitToFaceDistance to receive the distance of closest hit to beam spot
+         */
+        void GetSelectedCaloHits(const pandora::CaloHitList &inputCaloHitList, pandora::CaloHitList &outputCaloHitList,
+            double &closestHitToFaceDistance) const;
+
+        /** 
+         *  @brief  Find the intercepts of a line with the protoDUNE detector
+         *
+         *  @param  a0 a point on the line in question
+         *  @param  majorAxis the direction of the line in question
+         *  @param  interceptOne to receive the first intersection between line and protoDUNE detector 
+         *  @param  interceptTwo to receive the second intersection between line and protoDUNE detector 
+         */
+        void GetTPCIntercepts(const pandora::CartesianVector &a0, const pandora::CartesianVector &majorAxis,
+            pandora::CartesianVector &interceptOne, pandora::CartesianVector &interceptTwo) const;
+
+        /**
+         *  @brief  Check if a given 3D spacepoint is inside the global TPC volume
+         *
+         *  @param  spacePoint
+         */
+        bool IsContained(const pandora::CartesianVector &spacePoint) const;
+
+        bool                            m_isAvailable;               ///< Is the feature vector available
+        const SliceFeatureParamters    *m_pSliceFeatureParamters;    ///< Geometry information block
```

Comment:
Maybe not on the heap?

Reply:
The SliceFeatureParameters information block is needed to construct the SliceFeature class, therefore, instead of making many copies of the SliceFeatureParameters class, I thought it more sensible to have one and to pass pointers to that const class.  If it were on stack then there'd be lots of copies of the same object floating around, which I was trying to avoid.  Does that make sense?